src/V3Sched.cpp

```diff
@@ -822,71 +857,97 @@ void createEval(AstNetlist* netlistp,  //
                   // Resume triggered timing schedulers
                   if (AstCCall* const resumep = timingKit.createResume(netlistp)) {
                       resultp = AstNode::addNext(resultp, resumep->makeStmt());
                   }
-
                   // Invoke body function
                   {
-                      AstCCall* const callp = new AstCCall{flp, actFuncp};
+                      AstCCall* const callp = new AstCCall{flp, actKit.m_funcp};
                       callp->dtypeSetVoid();
-                      return AstNode::addNext(resultp, callp->makeStmt());
+                      resultp = AstNode::addNext(resultp, callp->makeStmt());
                   }
 
                   return resultp;
               })
               .second;
 
     // Create the NBA eval loop. This uses the Active eval loop in the trigger section.
-    AstNodeStmt* const nbaEvalLoopp
+    AstNodeStmt* topEvalLoopp
         = makeEvalLoop(
-              netlistp, "nba", "NBA", nbaTrigsp, nbaDumpp,
+              netlistp, "nba", "NBA", nbaKit.m_vscp, nbaKit.m_dumpp,
               [&]() {  // Trigger
-                  AstNodeStmt* resultp = nullptr;
-
                   // Reset NBA triggers
-                  {
-                      AstVarRef* const refp = new AstVarRef{flp, nbaTrigsp, VAccess::WRITE};
-                      AstCMethodHard* const callp = new AstCMethodHard{flp, refp, "clear"};
-                      callp->dtypeSetVoid();
-                      resultp = AstNode::addNext(resultp, callp->makeStmt());
-                  }
-
+                  AstNodeStmt* resultp = createTriggerClearCall(flp, nbaKit.m_vscp);
                   // Run the Active eval loop
-                  return AstNode::addNext(resultp, activeEvalLoopp);
+                  resultp = AstNode::addNext(resultp, activeEvalLoopp);
+                  return resultp;
               },
               [&]() {  // Body
-                  AstCCall* const callp = new AstCCall{flp, nbaFuncp};
+                  AstCCall* const callp = new AstCCall{flp, nbaKit.m_funcp};
                   callp->dtypeSetVoid();
-                  return callp->makeStmt();
+                  AstNodeStmt* resultp = callp->makeStmt();
+                  // Latch the NBA trigger flags under the following region's trigger flags
+                  AstVarScope* const nextVscp = obsKit.m_vscp ? obsKit.m_vscp : reactKit.m_vscp;
+                  if (nextVscp) {
+                      resultp = AstNode::addNext(
+                          resultp, createTriggerSetCall(flp, nextVscp, nbaKit.m_vscp));
+                  }
+                  return resultp;
               })
               .second;
 
-    // Add the NBA eval loop
-    funcp->addStmtsp(nbaEvalLoopp);
+    if (obsKit.m_funcp) {
+        // Create the Observed eval loop. This uses the NBA eval loop in the trigger section.
+        topEvalLoopp
+            = makeEvalLoop(
```

Comment:
Can this and the next one instead be normal functions called here?

Reply:
You mean the ones passed to `makeEvalLoop`?

Reply:
Yes

Reply:
Well, this is a structure introduced in V5, the `act` and `nba` loops are being constructed in a similar way. So if we want to pass nodes directly instead of the lambdas, this should be a pre-PR.